modules/fancymeshrenderer/src/calcnormals.cpp

```diff
@@ -29,60 +29,80 @@
 
 #include <modules/fancymeshrenderer/calcnormals.h>
 
+#include <inviwo/core/datastructures/buffer/buffer.h>
+#include <inviwo/core/datastructures/buffer/bufferram.h>
+#include <inviwo/core/datastructures/buffer/bufferramprecision.h>
+
 namespace inviwo {
-Mesh* CalcNormals::processMesh(const Mesh* const input, Mode mode) {
-    // short cut: pass through
-    Mesh* m = input->clone();
+
+namespace meshutil {
+using Mode = CalculateMeshNormalsMode;
+
+void calculateMeshNormals(Mesh& mesh, CalculateMeshNormalsMode mode) {
     if (mode == Mode::PassThrough) {
-        return m;
+        return;
     }
 
     // get input buffers
-    const BufferBase* bv = nullptr;
-    BufferBase* bn = nullptr;
-    for (int i = 0; i < m->getNumberOfBuffers(); ++i) {
-        if (m->getBuffers().at(i).first.type == BufferType::PositionAttrib) {
-            bv = m->getBuffers().at(i).second.get();
-        } else if (m->getBuffers().at(i).first.type == BufferType::NormalAttrib) {
-            bn = m->getBuffers().at(i).second.get();
-        }
+    auto [bv, bvLoc] = mesh.findBuffer(BufferType::PositionAttrib);
+    auto [bn, bnLoc] = mesh.findBuffer(BufferType::NormalAttrib);
+
+    BufferBase* asdf = const_cast<BufferBase*>(bn);
+
+    if (!bv) {
+        throw Exception("Input mesh has no position buffer",
+                        IVW_CONTEXT_CUSTOM("meshutil::calculateMeshNormals"));
     }
-    if (bv == nullptr || bn == nullptr) {
-        LogWarn("no position or normal buffer found in the mesh");
-        return m;
+
+    // Make sure we have a normal buffer of type vec3
+    if (!bn || bn->getDataFormat() != DataVec3Float32::get()) {
+        auto newBuf = std::make_shared<Buffer<vec3>>();
+        if (bn) {
+            mesh.replaceBuffer(bnLoc, {BufferType::NormalAttrib, bnLoc}, newBuf);
+        } else {
+            mesh.addBuffer(BufferType::NormalAttrib, newBuf);
+        }
+        bn = newBuf.get();
     }
-    auto vertices = bv->getRepresentation<BufferRAM>();
-    auto normals = bn->getEditableRepresentation<BufferRAM>();
 
-    // reset normals
-    for (int i = 0; i < normals->getSize(); ++i) {
-        normals->setFromDVec3(i, dvec3(0));
+    auto vertices = bv->getRepresentation<BufferRAM>();
```

Comment:
It is much better to construct every thing the other way around, vector->bufferRAM->buffer